### PR TITLE
UI Module

### PR DESCRIPTION
**Pull request by VengarvdWal**
UI Rendering branch that been refactored to use the new render systems. Font rendering is now also fixed. 

This also changes how each pass should have a output for the Framegraph. 

Currently the tonemapping pass outputs an image that is blitted onto the ui target, which does it's rendering, and then outputs a uiTarget and this is blitted to the swapchain on which the debug pass draws. 

Closes these decks
https://bubonic-brotherhood.codecks.io/decks/38-ui/card/1cv-refactor-ui-branch
https://bubonic-brotherhood.codecks.io/decks/38-ui/card/18x-ui-system
https://bubonic-brotherhood.codecks.io/milestones/run/40/card/1cq-ui-text-rendering
https://bubonic-brotherhood.codecks.io/milestones/run/40/card/18g-frame-graph-nodes-are-sometimes-in-the-incorrect-order

- [x] The render output should stay the same as before
- [x] In RenderDoc the passes are now correctly sorted 
- [x] Text is rendered
- [x] Buttons are rendered

### CODE REVIEW

File: engine/user_interface/private/ui_text.cpp [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    {
+        if (i != ' ')
+        {
+            // Check if the character exists in the font's character map
+            if (!font->characters.contains(i))
+            {
+                continue;
+            }
+
+            const auto& character = font->characters.at(i);
```
Comment (MuskettaMan) [inline engine/user_interface/private/ui_text.cpp:19]: unclear type

File: engine/user_interface/private/viewport.cpp [hunk 2]
```diff
@@ -0,0 +1,27 @@
+#include "viewport.hpp"
+#include "ui_element.hpp"
+
+void Viewport::Update(const InputManager& input) const
+{
+    for (const auto& i : _baseElements)
```
Comment (MuskettaMan) [inline engine/user_interface/private/viewport.cpp:6]: variable name

File: engine/user_interface/private/viewport.cpp [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    {
+        i->Update(input);
+    }
+}
+
+void Viewport::SubmitDrawInfo(std::vector<QuadDrawInfo>& drawList) const
+{
+    for (const auto& i : _baseElements)
```
Comment (MuskettaMan) [inline engine/user_interface/private/viewport.cpp:14]: variable name

File: engine/user_interface/public/ui_button.hpp [hunk 4]
```diff
@@ -0,0 +1,37 @@
+#pragma once
+#include "gpu_resources.hpp"
+#include "resource_manager.hpp"
+#include "ui_element.hpp"
+#include <functional>
+
+struct UIButton : public UIElement
```
Comment (MuskettaMan) [inline engine/user_interface/public/ui_button.hpp:7]: structs are inheriting from classes seems strange, i would keep it classes

File: engine/user_interface/public/ui_element.hpp [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+     */
+    NO_DISCARD const glm::vec2& GetRelativeLocation() const noexcept { return _relativeLocation; }
+    NO_DISCARD const glm::vec2& GetAbsouluteLocation() const noexcept { return _absoluteLocation; }
+
+    NO_DISCARD const glm::vec2& GetScale() const noexcept { return _scale; }
+    void SetScale(const glm::vec2& scale) noexcept { _scale = scale; }
+
+    virtual void SubmitDrawInfo(MAYBE_UNUSED std::vector<QuadDrawInfo>& drawList) const = 0;
+
+    virtual void Update(const InputManager& input)
```
Comment (MuskettaMan) [inline engine/user_interface/public/ui_element.hpp:51]: why are these definitions defined in the header file? They can be moved to the source file right

File: engine/user_interface/public/ui_element.hpp [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    {
+        for (auto& i : _children)
+            i->Update(input);
+    }
+
+    UIElement* AddChild(std::unique_ptr<UIElement> child)
```
Comment (MuskettaMan) [inline engine/user_interface/public/ui_element.hpp:57]: Sharing the backing pointers of unique ptrs, sounds contradicting, shouldnt we just use shared_ptr?

Comment (MuskettaMan) [reply engine/user_interface/public/ui_element.hpp:57]: or references if they wont be kept around

File: engine/user_interface/public/ui_element.hpp [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    {
+        if (_children.size() < _maxChildren && child != nullptr)
```
Comment (MuskettaMan) [inline engine/user_interface/public/ui_element.hpp:59]: unclear why there is max child count? isn't it just a vector backing this?

File: engine/user_interface/public/ui_module.hpp [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+class UIModule : public ModuleInterface
+{
+public:
+    UIModule() = default;
+    ~UIModule() override = default;
+
+    NON_COPYABLE(UIModule);
+    NON_MOVABLE(UIModule);
+
+    std::unique_ptr<Viewport> viewport;
```
Comment (MuskettaMan) [inline engine/user_interface/public/ui_module.hpp:14]: should this be public?

File: engine/user_interface/public/viewport.hpp [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+     * @param drawList Reference to the vector holding the QuadDrawInfo, which is most likely located in the uiPipeline.
+     */
+    void SubmitDrawInfo(std::vector<QuadDrawInfo>& drawList) const;
+
+    UIElement& AddElement(std::unique_ptr<UIElement> element);
+
+    NO_DISCARD std::vector<std::unique_ptr<UIElement>>& GetElements() { return _baseElements; }
+    NO_DISCARD const std::vector<std::unique_ptr<UIElement>>& GetElements() const { return _baseElements; }
+
+    glm::vec2 extend;
```
Comment (MuskettaMan) [inline engine/user_interface/public/viewport.hpp:32]: these should be private with getters

**Review by MuskettaMan — CHANGES_REQUESTED** [review]
Good change to the new module system! Made a few comments about inconsistencies in code style. Another thing I found strange was structs inheriting from classes, with the UI element. I think we want to keep them all classes, and that implies more encapsulation as well, that currently isnt in the structs (for button for example).

File: engine/CMakeLists.txt [hunk 10]
```diff
@@ -28,5 +28,6 @@ target_add_module(Engine renderer Renderer)
 target_add_module(Engine physics Physics)
 target_add_module(Engine steam Steam)
 target_add_module(Engine audio Audio)
+target_add_module(Engine user_interface UserInterface)
```
[the remark this answers is not in the document]
Comment (luuk234) [reply engine/CMakeLists.txt:31]: this is the folder name

[the remark this answers is not in the document]
Comment (luuk234) [reply engine/CMakeLists.txt:31]: I assume we use the same naming convention for folders as files?

File: engine/renderer/private/gpu_resources.cpp [hunk 11]
```diff
@@ -75,6 +77,36 @@ Sampler& Sampler::operator=(Sampler&& other) noexcept
...
+void CPUImage::ExtractDataFromPNG(std::string_view path)
+{
+    int width;
+    int height;
+    int nrChannels; // dummy
+
+    std::byte* data = reinterpret_cast<std::byte*>(stbi_load(std::string(path).c_str(), &width, &height, &nrChannels,
+        4));
+
+    if (data == nullptr)
```
[the remark this answers is not in the document]
Comment (luuk234) [reply engine/renderer/private/gpu_resources.cpp:89]: clang tidy does but have we ever made a formal decision on this? i see if in a lot of other places in the codebase

File: engine/renderer/public/renderer.hpp [hunk 12]
```diff
@@ -55,13 +58,14 @@ class Renderer
     // TODO: Unavoidable currently, this needs to become a module
     ApplicationModule& _application;
     std::shared_ptr<ECS> _ecs;
-
+    Viewport& _viewport;
```
[the remark this answers is not in the document]
Comment (luuk234) [reply engine/renderer/public/renderer.hpp:61]: I would prefer clear ownership.

**Review by MuskettaMan — APPROVED** [review]

**Pull request merged**